geps/gep-2907/index.md

```diff
@@ -0,0 +1,150 @@
+
+### "Frontend" and "Backend"
+A guiding principle in this naming is to use consistent naming for “Upstream”
+(1+2) and “Downstream” (3+4), similar to Envoy. To avoid the confusion with what
+is upstream and downstream, Gateway API will use “Frontend” and “Backend”.
+
+1. Frontend: The entity connecting to a Gateway, typically a client application
+   and/or web browser.
+2. Backend: The entity a Gateway is routing traffic to, typically the endpoints
+   behind a Service.
+
+There are essentially 4 different segments that could be relevant to TLS
+configuration in Gateway API:
+
+```mermaid
+flowchart LR
+    Frontend -.->|1. Client Cert| Gateway
+    Gateway -.->|2. Server Cert| Frontend
+    Gateway -.->|3. Client Cert| Backend
+    Backend -.->|4. Server Cert| Gateway
+```
+
+The above diagram depicts these four segments as edges in a graph.
+
+## Proposed Segments
+Note that this does not represent any form of commitment that any of these
+fields or concepts will exist within Gateway API. If or when they do, we propose
+the following naming structure:
+
+### 1. Validate Client Certificate provided by Frontend
+
+| Proposed Placement | Name | Status |
+|-|-|-|
+| Gateway Listener | `Listener.TLS.FrontendValidation` | Proposed |
+
+#### Rationale
+Use FrontendValidation to leave room for concepts like trust anchors and trust
+domains. Anything not strictly tied to validation would belong to Listener.TLS
+which is now reserved exclusively for “Frontend” TLS configuration (1+2).
+
+#### Why Not Frontend.Validation?
+Part of the idea behind this GEP is that Listener.TLS is really entirely focused
+on 1+2 - the bits of TLS between frontend and Gateway. That means that if we
+were to add any additional TLS configuration in the Listener.TLS struct, it
+would tied to that limited scope, and thus we don't really need a separate
+Frontend struct.
+
+One could make an argument that Listener.TLS.Validation should be the field name
+here to avoid any ambiguity, but in this specific context, we think that it's
+probably helpful to specifically spell out frontend.
+
+
+### 2. Configure TLS Termination, including Server Certificate
+
+| Proposed Placement | Name | Status |
+|-|-|-|
+| Gateway Listener | `Listener.TLS` | GA |
+
+#### Rationale
+This is already finalized in the API and so we're stuck with this name. In
+hindsight a name that was more clearly tied to frontend TLS would have been
+ideal here.
+
+
+### 3. Configure Client Certificate that Gateway should use to connect to Backend
+
+| Proposed Placement | Name | Status |
+|-|-|-|
+| Gateway | `Gateway.Spec.BackendTLS.ClientCertificateRef` | Not Proposed |
+| BackendTLSPolicy | `BackendTLSPolicy.Spec.ClientCertificateRef` | Not Proposed |
+
+#### Rationale
+It's not yet obvious which of the above options are preferable, but a case could
+be made for either or both. If we add a `BackendTLS` struct to Gateway it would
+leave room for concepts like TLS version, ALPN, cipher suites, etc.
```

Comment:
@robscott I feel we shouldn't add certain backend configuration that the app developer should have access to.  This part should be left TBD if it's not needed to unblock GEP-91.  We should not have a proposal that includes the indecision of "either or both" right from the beginning.

I mentioned earlier:

> Remembering the discussions around why the content of BackendTLSPolicy wasn't going to be added directly to Gateway, one reason was because BackendTLS should be accessible to the app developer role, and the burden of app or service-level certificate management shouldn't be placed on the cluster operator. How do we reconcile those discussions with this proposal? Add a new role, like application operator or similar?


>How about considering a new policy attachment called FrontendTLSPolicy, which would be available to cluster operators, but maybe not app developers. It would be similar to BackendTLSPolicy in many ways, but perhaps targetRefs could be listeners on the gateway? Would it work for @arkodg's use cases ?

Reply:
Good question, I think there are several reasons BackendTLSPolicy exists as a separate policy resource:

1. Some parts of the config, such as the SNI that should be used to connect to the Service/backend only make sense when attached per Service - it's very unlikely that you'd want to use the same SNI to connect to all backends from a Gateway.
2. The Service API is overloaded and very slow to change, Gateway API has taken the approach of exposing config at that level via policy resources that are tied to specific personas (why Session Persistence config and Backend TLS are in different resources).

I don't think either of those reasons really apply to frontend TLS. Instead, frontend TLS could theoretically be configured either for an entire Gateway, or a specific Gateway listener. Given that many implementations already support distinguishing this config per SNI, it seems to make sense to start with listener level attachment. I think that the persona that would be responsible for a Gateway is not sufficiently different than the persona that would be responsible for frontend TLS, so I tend to think that the current proposal is the best option available to us.

I'll work on clarifying all of that in the GEP.

Reply:
Added a new section "Why Not FrontendTLSPolicy?" to address this.